fetcher/fetcher.go

```diff
@@ -0,0 +1,153 @@
+package fetcher
+
+import (
+	"covid/database"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"time"
+)
+
+type Covid19APIResponse struct {
+	Country     string `json:"Country"`
+	CountryCode string `json:"CountryCode"`
+	Province    string `json:"Province"`
+	Cases       int    `json:"Cases"`
+	Status      string `json:"Status"`
+	Date        string `json:"Date"`
+}
+
+var counter int = 0
+
+func StartFetchingRoutine(db *sql.DB, updateInterval time.Duration) {
+	go func() {
+		FetchAndUpdateData(db) // Call this function immediately for the initial data fetch.
+		for range time.Tick(updateInterval) {
+			FetchAndUpdateData(db)
+		}
+	}()
+}
+
+func FetchAndUpdateData(db *sql.DB) error {
+	d := database.NewDB(db)
+	countries, err := d.GetCountries(nil, nil, nil, nil)
+	if err != nil {
+		log.Printf("Error fetching country list: %v", err)
+		return err
+	}
+
+	for _, country := range countries {
+		confirmedData, err := FetchDailyDataForCountry(country.Name, "confirmed")
+		if err != nil {
+			log.Printf("Error fetching confirmed data for country %s: %v", country.Name, err)
+			continue
+		}
+
+		deathsData, err := FetchDailyDataForCountry(country.Name, "deaths")
+		if err != nil {
+			log.Printf("Error fetching deaths data for country %s: %v", country.Name, err)
+			continue
+		}
+
+		recoveredData, err := FetchDailyDataForCountry(country.Name, "recovered")
+		if err != nil {
+			log.Printf("Error fetching recovered data for country %s: %v", country.Name, err)
+			continue
+		}
+
+		if err := UpdateCountryData(db, country.Name, country.ID, confirmedData, deathsData, recoveredData); err != nil {
+			log.Printf("Error updating country data for %s: %v", country.Name, err)
+			continue
+		}
+	}
+	return nil
+}
+
+func FetchDailyDataForCountry(countryName string, status string) ([]Covid19APIResponse, error) {
+	url := fmt.Sprintf("https://api.covid19api.com/dayone/country/%s/status/%s", countryName, status)
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	err = handleRateLimitingError(resp, countryName, status)
+	if err != nil {
+		return nil, err
+	}
+
+	var data []Covid19APIResponse
+	if err := json.Unmarshal(body, &data); err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func handleRateLimitingError(resp *http.Response, countryName string, status string) error {

```

Comment:
it has a side effect of calling the `FetchDailyDataForCountry` function recursively. This may cause issues with the stack as retries increase. I would suggest refactoring it to return a boolean indicating whether a retry is needed and handling the retry in the `FetchDailyDataForCountry` function itself.